File: 238-product-of-array-except-self/product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    let output = []


    let curr = 1
    for(let i = 0; i < nums.length; i++) {
        output.push(curr)
        // output[i] = curr
        curr = curr * nums[i]
    }

    curr = 1
    console.log(output)
    for(let i = nums.length - 1; i >= 0; i--) {
        output[i] = output[i] * curr
        curr = curr * nums[i]
    }

    return output
};

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ans = [1] * (len(nums))

        prefix = 1
        for i in range(len(nums)):
            ans[i] = prefix
            prefix *= nums[i]
        
        suff = 1
        for i in range(len(nums) -1, -1, -1):
            ans[i] = ans[i] * suff
            suff = suff * nums[i]
        
        return ans
    
    
        